src/ethereum/frontier/vm/instructions/system.py

```diff
@@ -260,3 +265,38 @@ def callcode(evm: Evm) -> None:
         child_evm.output[:actual_output_size],
     )
     evm.gas_left += child_evm.gas_left
+    evm.refund_counter += child_evm.refund_counter
+    evm.accounts_to_delete.update(child_evm.accounts_to_delete)
+
+
+def selfdestruct(evm: Evm) -> None:
+    """
+    Halt execution and register account for later deletion.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    beneficiary = to_address(pop(evm.stack))
+    originator = evm.message.current_target
+    beneficiary_balance = get_account(evm.env.state, beneficiary).balance
+    originator_balance = get_account(evm.env.state, originator).balance
+
+    # First Transfer to beneficiary
+    set_account_balance(
+        evm.env.state, beneficiary, beneficiary_balance + originator_balance
+    )
+    # Next, Zero the balance of the address being deleted (must come after
+    # sending to beneficiary in case the contract named itself as the
+    # beneficiary).
+    set_account_balance(evm.env.state, originator, U256(0))
+
+    # gas refund
+    evm.refund_counter += REFUND_SELF_DESTRUCT
+
+    # register account for deletion
+    evm.accounts_to_delete.add(beneficiary)
```

Comment:
This is incorrect. I should register the `originator` for delete and not the beneficiary. 
Sadly no test caught this. I'm looking for tests at the moment that will catch this bug.